### DISCUSSION TITLE
Support Eagle-3 Speculative Decoding in llama.cpp

### DISCUSSION THREAD

**Discussion by ruixiang63** [ggml-org/llama.cpp#15902] [Ideas]
Eagle-3 is currently the SOTA algorithm for speculative decoding, as demonstrated by [Spec-bench](https://github.com/hemingkx/Spec-Bench/blob/main/Leaderboard.md) and the [Eagle-3 paper](https://arxiv.org/pdf/2503.01840). However, llama.cpp does not yet support Eagle-3 while other major LLM inference frameworks, such as [TRT-LLM](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/eagle/README.md), [vLLM](https://developers.redhat.com/articles/2025/07/01/fly-eagle3-fly-faster-inference-vllm-speculative-decoding#), and [SGLang](https://github.com/sgl-project/sglang/pull/4247), already provide support for Eagle-3, resulting in around 2-2.5x performance boost compared to native autoregressive decoding. 

Furthermore, there are several PR & issues  for implementing Eagle-3 in llama.cpp: #13908, #15305

Many models with Eagle-3 checkpoints are already available on Hugging Face ([link](https://huggingface.co/models?sort=trending&search=eagle-3)), and users can also fine-tune their own Eagle-3 checkpoints using [TensorRT-Model-Optimizer](https://nvidia.github.io/TensorRT-Model-Optimizer/guides/5_speculative_decoding.html). 

Based on the above, I see a significant need to implement Eagle-3 in llama.cpp to potentially make LLM inference faster and llama.cpp more competitive. Therefore, I would like to initiate a discussion with the llama.cpp team to align on the goals and implementation.

To implement Eagle-3 in llama.cpp, several components need to be addressed (this may not be 100% accurate, and I am happy to receive any feedback on it):
* Eagle-3 Draft Model Checkpoint Support:
    * Convert Eagle-3 Draft PyTorch Checkpoints to GGUF Format
    * Eagle-3 Draft Model Architecture Support
    * Feature Extraction from Target Model Forward Pass: Hidden states after the first, middle, and last decoding layers.
* Draft Tokens Sampling & Parallel Verification with Target Model:
    * context-dependent dynamic draft tree structure support
    * Acceptance & Rejection Strategy for Draft Tokens ([Speculative Sampling](https://arxiv.org/pdf/2211.17192))
* Repeat above until max-token-len or EOS token

Workflow: During inference, we need to record low, middle, and high-level features (hidden states after first, middle, and last decoding layers) in the forward pass of the target model. After that, we combine those hidden states and token embedding and feed it to the speculative layer. The speculative layer then generates a sequence of draft tokens autoregressively for parallel verification by the target model. 

Since the Eagle-3 checkpoint is model-specific, I propose to start with [llama3](https://huggingface.co/yuhuili/EAGLE3-LLaMA3.1-Instruct-8B). I appreciate your feedbacks on it. 

Updates:
Eagle3 support now is available in this PR: https://github.com/ggml-org/llama.cpp/pull/18039

**Comment by lippman1125** [comment]
absolutely necessary!

**Comment by ggerganov** [comment]
> context-dependent dynamic draft tree structure support

How important do you think this step is? My experience from some time ago with tree-based speculative decoding is that it does not bring much benefit to basic, linear speculative sampling. And at the same time it complicates the logic quite a bit.

> After that, we combine those hidden states and token embedding and feed it to the speculative layer.

Does the speculative layer maintain any additional state? Or is it purely a function of the 3 hidden states and the last token embedding?

**Comment by ruixiang63** [comment]
> Does the speculative layer maintain any additional state? Or is it purely a function of the 3 hidden states and the last token embedding?

Yes, the draft model / speculative layer consists of two FC layers, one decoder layer, and one LM head. During the training stage, the parameters of these components will be fine-tuned, resulting in a draft model checkpoint. This is why we need to convert the draft model checkpoint into GGUF format and add architecture support in the llama.cpp as well. 
<img width="522" height="783" alt="image" src="https://github.com/user-attachments/assets/16bc6676-ccab-4bf8-be82-ba5dd2ee2902" />


> How important do you think this step is? My experience from some time ago with tree-based speculative decoding is that it does not bring much benefit to basic, linear speculative sampling. And at the same time it complicates the logic quite a bit.

I agree it will complicate the logic a lot. In theory, tree-based decoding allows for the verification of more candidate tokens from a candidates tree simultaneously compared to linear decoding, which could improve the accept rate of draft tokens. Linear sampling includes only a single candidate token to verify per step, resulting in suboptimal speculative accuracy.
 
I checked PR #3624. It uses static expansion of the token tree, sampling `n` candidate tokens at each decoding step to form the tree node. Eagle-3 supports context-dependent dynamic tree sampling, which approximates the acceptance
rate using the confidence of the draft model and dynamically generates the draft tree based on this, performing **pruning of the draft tree** at the end of the drafting stage (details in [eagle-2 paper](https://arxiv.org/pdf/2406.16858)). 

Since the benefits of tree-based speculative decoding depend on model quality, Eagle-3 utilizes the tree attention mask during draft model training, which I believe aligns with tree-based sampling during inference. See the following image.

In summary, tree-based sampling is necessary for Eagle-3, whereas tree-based candidate verification is not mandatory. Static tree sampling is acceptable for Eagle-3, but according to the paper, context-dependent dynamic tree sampling works better. Based on this, I propose that the initial implementation should use **static tree sampling combined with batched candidate sequence verification** for simplicity. We can then iterate and improve from there.
@ggerganov 

<img width="1118" height="675" alt="image" src="https://github.com/user-attachments/assets/114dcec5-317e-4741-b21a-ef5c738e347a" />

**Reply by ggerganov** [reply]
Based on this, I see the following strategy for implementation:

- Convert Eagle checkpoints to GGUF
- Add support for loading Eagle GGUFs similar to all other models (i.e. a separate architecture in `llama-model`)
- From the diagram, it looks like the Eagle architecture has to be implemented as Encoder-Decoder in order to support the "FC Layer" for the extracted intermediate embeddings of the target model. Basically, the Encoder will apply a single FC to an input batch of `h,m,l` embeddings in order to obtain the `g` embeddings (i.e. use `llama_encode`). The `g` embeddings will then be used in the Decoder to process an input batch of `e` tokens (i.e. use `llama_decode`).
- I think the computed `g` embeddings from the Encoder, together with the accumulated `a` embeddings after each Decode step should probably be stored into a new auxilary struct in the `llama_context`, similar to the existing `llama_cross`. Generally, `llama_cross` is in a very rough state at the moment and likely will be redesigned in the future in some way, but I think for now, we can follow the existing implementation.
- It is not clear from the diagram, how big the context of the Eagle can become? Does it keep track of the last few tokens, or does it grow as big as the context of the Target model?
- For the Target model, we need to introduce a mechanism to extract `h,m,l` embeddings. This can be achieved in a similar way as described in https://github.com/ggml-org/llama.cpp/pull/15225#issuecomment-3182616064

Let me know if this makes sense or if something is unclear.

**Reply by ruixiang63** [reply]
Thank you very much for the valuable feedback.

> From the diagram, it looks like the Eagle architecture has to be implemented as Encoder-Decoder in order to support the "FC Layer" for the extracted intermediate embeddings of the target model. Basically, the Encoder will apply a single FC to an input batch of h,m,l embeddings in order to obtain the g embeddings (i.e. use llama_encode). The g embeddings will then be used in the Decoder to process an input batch of `e` tokens (i.e. use llama_decode).

This is almost correct. The missing part is `e`, which represents the embeddings of the generated tokens. The draft-eagle model shares the same embedding layer as the target model. I checked an example `yuhuili/EAGLE3-LLaMA3.1-Instruct-8B` checkpoint, and it seems the checkpoint already has the embedding layer, so we don't need to worry about reusing the embedding layer from the target model. Example of eagle-3 model structure of `yuhuili/EAGLE3-LLaMA3.1-Instruct-8B` 
```
LlamaModel(
  (embed_tokens): Embedding(128256, 4096, padding_idx=0)
  (layers): ModuleList(
    (0): LlamaDecoderLayer(
      (self_attn): LlamaAttention(
        (q_proj): Linear(in_features=4096, out_features=4096, bias=False)
        (k_proj): Linear(in_features=4096, out_features=1024, bias=False)
        (v_proj): Linear(in_features=4096, out_features=1024, bias=False)
        (o_proj): Linear(in_features=4096, out_features=4096, bias=False)
      )
      (mlp): LlamaMLP(
        (gate_proj): Linear(in_features=4096, out_features=14336, bias=False)
        (up_proj): Linear(in_features=4096, out_features=14336, bias=False)
        (down_proj): Linear(in_features=14336, out_features=4096, bias=False)
        (act_fn): SiLU()
      )
      (input_layernorm): LlamaRMSNorm((4096,), eps=1e-05)
      (post_attention_layernorm): LlamaRMSNorm((4096,), eps=1e-05)
    )
  )
  (norm): LlamaRMSNorm((4096,), eps=1e-05)
  (rotary_emb): LlamaRotaryEmbedding()
)
```

> It is not clear from the diagram, how big the context of the Eagle can become? Does it keep track of the last few tokens, or does it grow as big as the context of the Target model?

I don't fully understand this question. If you are referring to the context regarding the KV cache, I would say the context is as large as the target model. (*With tree-based decoding, it can become even larger*) Both the eagle model and the target model maintain seperate KV caches. The target model only maintains the KV cache for accepted or correct tokens, while the eagle model maintains the KV cache for all candidate tokens, including those that might be rejected by the target model. Therefore, we also need a KV cache eviction mechanism for the eagle model.

I have one question is about the draft tokens generation and verification implementation. As I proposed before, the initial implementation should use static tree sampling combined with batched candidate sequence verification for simplicity. Do you have any suggestions regarding this? Do we need to implement some features inside llama.cpp to support this, or can the current implementation in llama.cpp support it? @ggerganov

**Reply by ggerganov** [reply]
By KV cache of the Eagle model do you mean the set of `g` embeddings? Want to clear this up since there is no attention in the Eagle model, so "KV cache" does not make much sense in that case.

If we need to store the full set of `g` embeddings for the entire context (as opposed to the last few), then we probably have to implement a new `llama_memory` type to store this data.

Edit: Ok, I now see there is actually an attention layer in the Eagle model.

So for the KV cache of the Eagle model, we can use the existing KV cache memory module as it is. But it is still not clear to me how to store the `g` embeddings.

**Reply by ggerganov** [reply]
> I have one question is about the draft tokens generation and verification implementation. As I proposed before, the initial implementation should use static tree sampling combined with batched candidate sequence verification for simplicity. Do you have any suggestions regarding this? Do we need to implement some features inside llama.cpp to support this, or can the current implementation in llama.cpp support it?

Generally, any kind of sampling tree is supported, as demonstrated in #3624. Multi-sequence sampling trees require the unified KV cache to be enabled.

> I propose that the initial implementation should use static tree sampling combined with batched candidate sequence verification for simplicity. 

It would be even simpler to start with a basic linear draft of a few tokens with the Eagle and verify those in a batch. For example, at token `N`, you can draft tokens `N + 1` and `N + 2` and verify those. This way you don't need to construct a tree and this also works with split KV cache.

**Reply by ruixiang63** [reply]
> So for the KV cache of the Eagle model, we can use the existing KV cache memory module as it is. But it is still not clear to me how to store the g embeddings.

We don't need to store the `g` embeddings as they are only computed once and then concatenated with `e` to obtain the KV cache in Eagle model. As long as we have the KV cache for the Eagle model, storing the `g` embeddings is unnecessary.
At the end of each verification stage, a new token will be generated from the target model, resulting in the generation of `h`, `m`, and `l`. In this case, we will calculate the `g` embedding on the fly, no need to store `g` as well.

> It would be even simpler to start with a basic linear draft of a few tokens with the Eagle and verify those in a batch. For example, at token N, you can draft tokens N + 1 and N + 2 and verify those. This way you don't need to construct a tree and this also works with split KV cache.

Yes, thanks for the suggestion. It's worth starting with simplicity.

**Reply by ruixiang63** [reply]
The overall implementation mini plan sounds clear to me and is a good starting point. I will prepare to begin the implementation. If you have any questions or concerns, please let me know. I believe we will have more questions during the actual implementation, and we can always continue our discussion as needed. :) 
Thanks! @ggerganov

**Reply by ggerganov** [reply]
> We don't need to store the g embeddings as they are only computed once and then concatenated with e to obtain the KV cache in Eagle model.

Ok, this is great.

> In this case, we will calculate the g embedding on the fly, no need to store g as well.

If you can clarify how this is performed, it would help me to get a clearer picture. On the diagram, we apply an "FC Layer" to the `h,m,l` embeddings. What does this layer consist of? (i.e. what are the tensors, shapes and operators involved)

> The overall implementation mini plan sounds clear to me and is a good starting point.

Yes, happy to help if you get stuck somewhere - let me know.

**Reply by ggerganov** [reply]
I think the `speculative-simple` example can be a good starting point for experimentation. Unless I am missing something, the example can be adapted for Eagle by implementing an alternative `common_speculative_gen_draft` that works with Eagle models instead of basic draft model.

**Reply by ruixiang63** [reply]
> If you can clarify how this is performed, it would help me to get a clearer picture. On the diagram, we apply an "FC Layer" to the h,m,l embeddings. What does this layer consist of? (i.e. what are the tensors, shapes and operators involved)

We start with an example: from the target model's forward pass, we obtain `aux_hidden_states`, which is a list of hidden states `[l, m, h]`. The selection of each hidden state is heuristic, as described in the original paper, they use the indices `[2, num_layers // 2, num_layers - 3]` from the target model layers. The shape of each `[h, m, l]` is `[batch_size, seq_len, hidden_dim]`, and for this example, we use `[1, 32, 4096]`.
The FC Layer consists of a single fully connected MLP layer, e.g., `fc = torch.nn.Linear(hidden_dim * 3, hidden_dim)`, which takes the concatenated `[h, m, l]` with shape `[1, 32, 4096 * 3]` as input. The output size of the FC Layer is `[1, 32, 4096]`, which is `g`. 
Once we obtain `g`, it will be concatenated with `e` and fed into the draft model. 
The FC (fully connected) layer in the diagram usually refers to a single MLP linear layer.

> I think the speculative-simple example can be a good starting point for experimentation. Unless I am missing something, the example can be adapted for Eagle by implementing an alternative common_speculative_gen_draft that works with Eagle models instead of basic draft model.

Great suggestion. Thanks!
